runelite-client/src/main/java/net/runelite/client/plugins/antidrag/AntiDragPlugin.java

```diff
@@ -84,31 +110,86 @@ public void keyTyped(KeyEvent e)
 	@Override
 	public void keyPressed(KeyEvent e)
 	{
-		if (e.getKeyCode() == KeyEvent.VK_SHIFT)
+		if (config.onShiftOnly() && e.getKeyCode() == KeyEvent.VK_SHIFT)
 		{
-			final int delay = config.dragDelay();
-			client.setInventoryDragDelay(delay);
-			setBankDragDelay(delay);
+			setDragDelay();
 		}
 	}
 
 	@Override
 	public void keyReleased(KeyEvent e)
 	{
-		if (e.getKeyCode() == KeyEvent.VK_SHIFT)
+		if (e.getKeyCode() == KeyEvent.VK_SHIFT && config.onShiftOnly())
+		{
+			resetDragDelay();
+		}
+	}
+
+	@Subscribe
+	public void onConfigChanged(ConfigChanged event)
+	{
+		if (event.getGroup().equals(CONFIG_GROUP))
+		{
+			clientThread.invoke(() ->
+			{
+				if (config.onShiftOnly() || inPvp())
+				{
+					resetDragDelay();
+				}
+				else
+				{
+					setDragDelay();
+				}
+			});
+		}
+	}
+
+	@Subscribe
+	public void onVarbitChanged(VarbitChanged varbitChanged)
+	{
+		boolean currentStatus = inPvp();
+
+		if (currentStatus != lastInPvp)
 		{
-			client.setInventoryDragDelay(DEFAULT_DELAY);
-			setBankDragDelay(DEFAULT_DELAY);
+			if (!currentStatus && !config.onShiftOnly())
+			{
+				setDragDelay();
+			}
+			else
+			{
+				resetDragDelay();
+			}
+
+			lastInPvp = currentStatus;
 		}
+
 	}
 
 	@Subscribe
 	public void onFocusChanged(FocusChanged focusChanged)
 	{
 		if (!focusChanged.isFocused())
 		{
-			client.setInventoryDragDelay(DEFAULT_DELAY);
-			setBankDragDelay(DEFAULT_DELAY);
+			resetDragDelay();
+		}
+		else
+		{
+			clientThread.invokeLater(() ->
+			{
+				if (!config.onShiftOnly() && !inPvp())
+				{
+					setDragDelay();
```

Comment:
Only the set (and probably the above resetDragDelay()) needs to be on client thread, the rest is fine off.